delivery_carrier_label_ups/models/stock_picking.py

```diff
@@ -0,0 +1,262 @@
+            ),
+            PackageServiceOptions="",
+        )
+
+    def _ups_label_data(self, package):
+        """Return a dict that can be passed to the label endpoint of the UPS API"""
+        self.ensure_one()
+        account = self._ups_auth_account()
+        return dict(
+            LabelRecoveryRequest=dict(
+                LabelSpecification=dict(
+                    LabelImageFormat=dict(Code=account.file_format or "PDF",),
+                ),
+                # TODO we could add a Translate key here to translate to a few languages
+                TrackingNumber=package.parcel_tracking,
+            ),
+        )
+
+    def _ups_raise_error(self, api_result):
+        """If UPS' API returned an error, raise it"""
+        errors = api_result.get("response", {}).get("errors")
+        if errors:
+            raise UserError(
+                _("Sending %s to UPS: %s")
+                % (
+                    self.name,
+                    "\n".join("%(code)s %(message)s" % error for error in errors),
+                )
+            )
+
+    def _ups_request(
+        self,
+        request_type,
+        payload=None,
+        method="post",
+        url_parameters=None,
+        query_parameters=None,
+    ):
+        """Send a request to one of the UPS endpoints"""
+        self.ensure_one()
+        account = self._ups_auth_account()
+        if not account:
+            raise UserError(_("No UPS account found for this company"))
+        url = self.env["ir.config_parameter"].get_param(
+            "delivery_carrier_label_ups.%s_%s"
+            % (
+                request_type,
+                self.carrier_id.prod_environment and "production" or "test",
+            )
+        ) % (url_parameters or ())
+        headers = {
+            "AccessLicenseNumber": account.ups_access_license,
+            "Username": account.account,
+            "Password": account.password,
+            "transactionSrc": "Odoo (%s)" % self.env.cr.dbname,
+        }
+        return getattr(requests, method)(
+            url, json=payload, headers=headers, params=query_parameters
+        ).json()
+
+    def _ups_auth_account(self):
+        """Find a carrier.account matching our carrier"""
+        self.ensure_one()
+        return self.env["carrier.account"].search(
+            [
+                ("delivery_type", "=", "ups"),
+                "|",
+                ("company_id", "=", False),
+                ("company_id", "=", self.company_id.id),
+            ],
+            limit=1,
+            order="company_id desc",
+        )
+
+    def _ups_reference(self):
+        """Return a reference for this picking to be passed to UPS"""
+        self.ensure_one()
+        return self.name
+
+    def _ups_tracking_url(self, tracking_number):
+        """Return an URL for a tracking number"""
+        return "https://ups.com/WebTracking/track?trackingNumber=%s" % tracking_number
```

Comment:
Did you check if it is possible to provide a link containing all the tracking numbers from all the packages in the picking? I can't seem to get the format from our 8.0 module to work but maybe I'm doing something wrong or is there another way.
What documentation did you use for this module BTW?

Reply:
Seems we can just urlencode spaces in between them:
ups.com/WebTracking/track?trackingNumber=a+b+c

Reply:
Cool

Reply:
also done